IdentityCore/src/requests/MSIDSilentTokenRequest.m

```diff
@@ -235,129 +216,145 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
-                         NSError *msidError = nil;
-
-                         [self updateFamilyIdCacheWithServerError:error
-                                                       cacheError:&msidError];
-
-                         if (msidError)
-                         {
-                             MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, self.requestParameters, @"Failed to update familyID cache status with error %@", MSID_PII_LOG_MASKABLE(error));
-                         }
-
-                         MSIDBaseToken<MSIDRefreshableToken> *appRefreshToken = [self appRefreshTokenWithError:&msidError];
-
-                         if (msidError)
-                         {
-                             MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, self.requestParameters, @"Failed to retrieve multi resource refresh token with error %@", MSID_PII_LOG_MASKABLE(error));
-                             completionBlock(nil, msidError);
-                             return;
-                         }
-
-                         if (appRefreshToken && ![[familyRefreshToken refreshToken] isEqualToString:[appRefreshToken refreshToken]])
-                         {
-                             [self tryAppRefreshToken:appRefreshToken completionBlock:completionBlock];
-                             return;
-                         }
-
-                         NSError *interactionError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractionRequired, @"User interaction is required", error.msidOauthError, error.msidSubError, error, self.requestParameters.correlationId, nil, YES);
-                         completionBlock(nil, interactionError);
-                     }
-                     else
-                     {
-                         completionBlock(nil, error);
-                     }
-                 }
-                 else
-                 {
-                     completionBlock(result, nil);
-                 }
-             }];
+    BOOL isAppRefreshToken = tokenType == MSIDAppRefreshTokenType;
+    MSID_LOG_WITH_CTX_PII(MSIDLogLevelVerbose, self.requestParameters, @"Trying to acquire access token using %@ for clientId %@, authority %@, account %@", (!isAppRefreshToken ? @"App Refresh Token" : @"Family Refresh Token"), self.requestParameters.authority, self.requestParameters.clientId, self.requestParameters.accountIdentifier.maskedHomeAccountId);
+    
+    // When using ART or FRT, it will go through the same method below, and handle differently within the completion block
+    [self redeemAccessTokenWith:refreshToken
+                completionBlock:^(MSIDTokenResult * _Nullable result, NSError * _Nullable error) {
+        if (error)
+        {
+            if ([self isErrorRecoverableByUserInteraction:error])
+            {
+                if (isAppRefreshToken)
+                {
+                    [self handleErrorResponseForAppRefreshToken:refreshToken completionBlock:completionBlock];
```

Comment:
It looks like there's a possibility that completionBlock will be invoked twice with the new logic:

1. Inside handleErrorResponseForAppRefreshToken:
2. At the line 310

Could you double check that completionBlock wouldn't get invoked twice?

Reply:
Good catch, update the logic and should do a early return to stop current the code flow